Don't allocate result Tensors in out overloads: Reduction Ops

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53660 [experiment] AlwaysCall Backend Fallback
* #55321 Don't allocate result Tensors in out overloads: _linalg_solve_out_helper_cuda
* #53624 Boxing logic forwards arguments to stack
* #53556 Fix forwarding/move bug
* #53640 Don't allocate result Tensors in out overloads: at::kron_out()
* **#53218 Don't allocate result Tensors in out overloads: Reduction Ops**
* #53229 Fix Scalar formatting
* #53228 Fix boxing/unboxing for Scalar bool values

We have some operators that previously allowed you to pass in an undefined tensor to the out argument,
and then would go on to allocate that for you. This behavior is broken and doesn't work in JIT when things
are converted to/from IValues. Because of this, it blocks backend fallbacks because they force going
through IValue.

This PR removes that behavior and forces out arguments to be defined tensors.

It only looks at reduction ops for now, there's likely more PRs coming for other ops.

BC Breaking: This breaks BC of the C++ frontend API since those ops previously allowed calling with undefined tensors and that isn't allowed anymore.

Differential Revision: [D26795461](https://our.internmc.facebook.com/intern/diff/D26795461/)

ezyang commented on aten/src/ATen/native/ReduceOpsUtils.h:
```diff
@@ -148,11 +148,34 @@ static void allocate_reduction_result(
+    return src_type;
+  });
+
+  auto shape = DimVector(self.sizes());
```
Probably best to comment that this is a straight copy paste of the block above, or better yet, factor it into its own function.

bhosmer replied:
+1 to factoring

ezyang commented:
It looks like the duplicated logic doesn't perfectly reflect the original. If you want I can give you an accept (because I'm generally pro this PR) but it seems like there need to be substantive changes.

smessmer committed:
Update on "Don't allocate result Tensors in out overloads: Reduction Ops"


We have some operators that previously allowed you to pass in an undefined tensor to the out argument,
and then would go on to allocate that for you. This behavior is broken and doesn't work in JIT when things
are converted to/from IValues. Because of this, it blocks backend fallbacks because they force going
through IValue.

This PR removes that behavior and forces out arguments to be defined tensors.

It only looks at reduction ops for now, there's likely more PRs coming for other ops.

BC Breaking: This breaks BC since those ops previously allowed calling with undefined tensors and that isn't allowed anymore.

Differential Revision: [D26795461](https://our.internmc.facebook.com/intern/diff/D26795461/)

[ghstack-poisoned]

bhosmer reviewed (commented):
@ezyang not sure I understand your comment re doesn't perfectly reflect, can you elaborate?

ezyang commented:
> @ezyang not sure I understand your comment re doesn't perfectly reflect, can you elaborate?

Well, this PR is failing tests. So definitionally, something wrong has happened.

smessmer committed:
Update on "Don't allocate result Tensors in out overloads: Reduction Ops"


We have some operators that previously allowed you to pass in an undefined tensor to the out argument,
and then would go on to allocate that for you. This behavior is broken and doesn't work in JIT when things
are converted to/from IValues. Because of this, it blocks backend fallbacks because they force going
through IValue.

This PR removes that behavior and forces out arguments to be defined tensors.

It only looks at reduction ops for now, there's likely more PRs coming for other ops.

BC Breaking: This breaks BC since those ops previously allowed calling with undefined tensors and that isn't allowed anymore.

Differential Revision: [D26795461](https://our.internmc.facebook.com/intern/diff/D26795461/)

[ghstack-poisoned]

smessmer committed:
Update on "Don't allocate result Tensors in out overloads: Reduction Ops"


We have some operators that previously allowed you to pass in an undefined tensor to the out argument,
and then would go on to allocate that for you. This behavior is broken and doesn't work in JIT when things
are converted to/from IValues. Because of this, it blocks backend fallbacks because they force going
through IValue.

This PR removes that behavior and forces out arguments to be defined tensors.

It only looks at reduction ops for now, there's likely more PRs coming for other ops.

BC Breaking: This breaks BC since those ops previously allowed calling with undefined tensors and that isn't allowed anymore.

Differential Revision: [D26795461](https://our.internmc.facebook.com/intern/diff/D26795461/)

[ghstack-poisoned]

smessmer committed:
Update on "Don't allocate result Tensors in out overloads: Reduction Ops"


We have some operators that previously allowed you to pass in an undefined tensor to the out argument,
and then would go on to allocate that for you. This behavior is broken and doesn't work in JIT when things
are converted to/from IValues. Because of this, it blocks backend fallbacks because they force going
through IValue.

This PR removes that behavior and forces out arguments to be defined tensors.

It only looks at reduction ops for now, there's likely more PRs coming for other ops.

BC Breaking: This breaks BC since those ops previously allowed calling with undefined tensors and that isn't allowed anymore.

Differential Revision: [D26795461](https://our.internmc.facebook.com/intern/diff/D26795461/)

[ghstack-poisoned]

smessmer committed:
Update on "Don't allocate result Tensors in out overloads: Reduction Ops"


We have some operators that previously allowed you to pass in an undefined tensor to the out argument,
and then would go on to allocate that for you. This behavior is broken and doesn't work in JIT when things
are converted to/from IValues. Because of this, it blocks backend fallbacks because they force going
through IValue.

This PR removes that behavior and forces out arguments to be defined tensors.

It only looks at reduction ops for now, there's likely more PRs coming for other ops.

BC Breaking: This breaks BC of the C++ frontend API since those ops previously allowed calling with undefined tensors and that isn't allowed anymore.

Differential Revision: [D26795461](https://our.internmc.facebook.com/intern/diff/D26795461/)

[ghstack-poisoned]

smessmer committed:
Update on "Don't allocate result Tensors in out overloads: Reduction Ops"


We have some operators that previously allowed you to pass in an undefined tensor to the out argument,
and then would go on to allocate that for you. This behavior is broken and doesn't work in JIT when things
are converted to/from IValues. Because of this, it blocks backend fallbacks because they force going
through IValue.

This PR removes that behavior and forces out arguments to be defined tensors.

It only looks at reduction ops for now, there's likely more PRs coming for other ops.

BC Breaking: This breaks BC of the C++ frontend API since those ops previously allowed calling with undefined tensors and that isn't allowed anymore.

Differential Revision: [D26795461](https://our.internmc.facebook.com/intern/diff/D26795461/)

[ghstack-poisoned]

smessmer committed:
Update on "Don't allocate result Tensors in out overloads: Reduction Ops"


We have some operators that previously allowed you to pass in an undefined tensor to the out argument,
and then would go on to allocate that for you. This behavior is broken and doesn't work in JIT when things
are converted to/from IValues. Because of this, it blocks backend fallbacks because they force going
through IValue.

This PR removes that behavior and forces out arguments to be defined tensors.

It only looks at reduction ops for now, there's likely more PRs coming for other ops.

BC Breaking: This breaks BC of the C++ frontend API since those ops previously allowed calling with undefined tensors and that isn't allowed anymore.

Differential Revision: [D26795461](https://our.internmc.facebook.com/intern/diff/D26795461/)

[ghstack-poisoned]

smessmer commented:
tests are fixed. @ezyang @bhosmer PTAL

smessmer committed:
Update on "Don't allocate result Tensors in out overloads: Reduction Ops"


We have some operators that previously allowed you to pass in an undefined tensor to the out argument,
and then would go on to allocate that for you. This behavior is broken and doesn't work in JIT when things
are converted to/from IValues. Because of this, it blocks backend fallbacks because they force going
through IValue.

This PR removes that behavior and forces out arguments to be defined tensors.

It only looks at reduction ops for now, there's likely more PRs coming for other ops.

BC Breaking: This breaks BC of the C++ frontend API since those ops previously allowed calling with undefined tensors and that isn't allowed anymore.

Differential Revision: [D26795461](https://our.internmc.facebook.com/intern/diff/D26795461/)

[ghstack-poisoned]

smessmer commented:
nvm, rebasing broke them again

smessmer committed:
Update on "Don't allocate result Tensors in out overloads: Reduction Ops"


We have some operators that previously allowed you to pass in an undefined tensor to the out argument,
and then would go on to allocate that for you. This behavior is broken and doesn't work in JIT when things
are converted to/from IValues. Because of this, it blocks backend fallbacks because they force going
through IValue.

This PR removes that behavior and forces out arguments to be defined tensors.

It only looks at reduction ops for now, there's likely more PRs coming for other ops.

BC Breaking: This breaks BC of the C++ frontend API since those ops previously allowed calling with undefined tensors and that isn't allowed anymore.

Differential Revision: [D26795461](https://our.internmc.facebook.com/intern/diff/D26795461/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -592,13 +592,20 @@ Tensor& diff_out(const Tensor& self, int64_t n, int64_t dim, const c10::optional
-static ScalarType get_dtype(Tensor& result, const Tensor& self, optional<ScalarType> dtype,
-                            bool promote_integers=false) {
+inline ScalarType get_dtype(Tensor& result, optional<ScalarType> dtype) {
+  TORCH_CHECK(result.defined(), "Cannot create a new tensor inside a reduction op. You likely tried to call an operator with an out argument but the out argument was an undefined tensor.");
```
Sounds more like an INTERNAL_ASSERT to me :>

smessmer replied:
This can be triggered by user code that calls the C++ API with an undefined tensor

ezyang commented:
It would be helpful if you could describe the general strategy of what the refactor you did was. You've said the Why in the PR (great!) but the How would also be a great help.

ezyang commented on aten/src/ATen/native/ReduceOpsUtils.h:
```diff
@@ -134,10 +134,7 @@ static DimMask make_dim_mask(IntArrayRef dims, int64_t ndim) {
-    Tensor& result, const Tensor& self, DimMask mask, bool keepdim,
-    ScalarType dtype)
-{
+inline Tensor shape_from_dim_mask(const Tensor& self, DimMask mask, bool keepdim) {
```
this type looks wrong

smessmer commented:
> It would be helpful if you could describe the general strategy of what the refactor you did was. You've said the Why in the PR (great!) but the How would also be a great help.

I looked at how the out variants of the operators create the result tensor and replicated that logic in the functional variant before calling into the out variant. I added an assertion to the out variants making sure that the result tensor is defined.

smessmer committed:
Update on "Don't allocate result Tensors in out overloads: Reduction Ops"


We have some operators that previously allowed you to pass in an undefined tensor to the out argument,
and then would go on to allocate that for you. This behavior is broken and doesn't work in JIT when things
are converted to/from IValues. Because of this, it blocks backend fallbacks because they force going
through IValue.

This PR removes that behavior and forces out arguments to be defined tensors.

It only looks at reduction ops for now, there's likely more PRs coming for other ops.

BC Breaking: This breaks BC of the C++ frontend API since those ops previously allowed calling with undefined tensors and that isn't allowed anymore.

Differential Revision: [D26795461](https://our.internmc.facebook.com/intern/diff/D26795461/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/ReduceOpsUtils.h:
```diff
@@ -148,11 +145,24 @@ static void allocate_reduction_result(
+    ScalarType dtype)
+{
+  auto shape = shape_from_dim_mask(self, mask, keepdim);
+  TORCH_CHECK(result.defined(), "Cannot create a new tensor inside a reduction op. You likely tried to call an operator with an out argument but the out argument was an undefined tensor.");
```
`INTERNAL_ASSERT` me

smessmer replied:
This can be triggered by user code without a bug in PyTorch when they call an out overload from the C++ API with an undefined tensor. So this should stay a TORCH_CHECK.

ezyang commented on aten/src/ATen/native/ReduceOpsUtils.h:
```diff
@@ -148,11 +145,24 @@ static void allocate_reduction_result(
+  return shape;
+}
+
+static void allocate_reduction_result(
```
resize instead of allocate here feels more accurate

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -592,13 +592,20 @@ Tensor& diff_out(const Tensor& self, int64_t n, int64_t dim, const c10::optional
 
-static ScalarType get_dtype(Tensor& result, const Tensor& self, optional<ScalarType> dtype,
-                            bool promote_integers=false) {
+inline ScalarType get_dtype(Tensor& result, optional<ScalarType> dtype) {
```
The new set of overloads here seems incredibly error prone here. `get_dtype(result, dtype)` treats result correctly, but `get_dtype(result, dtype, false)` treats result as `self`. Better to name them differently...

smessmer committed:
Update on "Don't allocate result Tensors in out overloads: Reduction Ops"


We have some operators that previously allowed you to pass in an undefined tensor to the out argument,
and then would go on to allocate that for you. This behavior is broken and doesn't work in JIT when things
are converted to/from IValues. Because of this, it blocks backend fallbacks because they force going
through IValue.

This PR removes that behavior and forces out arguments to be defined tensors.

It only looks at reduction ops for now, there's likely more PRs coming for other ops.

BC Breaking: This breaks BC of the C++ frontend API since those ops previously allowed calling with undefined tensors and that isn't allowed anymore.

Differential Revision: [D26795461](https://our.internmc.facebook.com/intern/diff/D26795461/)

[ghstack-poisoned]

smessmer committed:
Update on "Don't allocate result Tensors in out overloads: Reduction Ops"


We have some operators that previously allowed you to pass in an undefined tensor to the out argument,
and then would go on to allocate that for you. This behavior is broken and doesn't work in JIT when things
are converted to/from IValues. Because of this, it blocks backend fallbacks because they force going
through IValue.

This PR removes that behavior and forces out arguments to be defined tensors.

It only looks at reduction ops for now, there's likely more PRs coming for other ops.

BC Breaking: This breaks BC of the C++ frontend API since those ops previously allowed calling with undefined tensors and that isn't allowed anymore.

Differential Revision: [D26795461](https://our.internmc.facebook.com/intern/diff/D26795461/)

[ghstack-poisoned]

smessmer committed:
Update on "Don't allocate result Tensors in out overloads: Reduction Ops"


We have some operators that previously allowed you to pass in an undefined tensor to the out argument,
and then would go on to allocate that for you. This behavior is broken and doesn't work in JIT when things
are converted to/from IValues. Because of this, it blocks backend fallbacks because they force going
through IValue.

This PR removes that behavior and forces out arguments to be defined tensors.

It only looks at reduction ops for now, there's likely more PRs coming for other ops.

BC Breaking: This breaks BC of the C++ frontend API since those ops previously allowed calling with undefined tensors and that isn't allowed anymore.

Differential Revision: [D26795461](https://our.internmc.facebook.com/intern/diff/D26795461/)

[ghstack-poisoned]

smessmer committed:
Update on "Don't allocate result Tensors in out overloads: Reduction Ops"


We have some operators that previously allowed you to pass in an undefined tensor to the out argument,
and then would go on to allocate that for you. This behavior is broken and doesn't work in JIT when things
are converted to/from IValues. Because of this, it blocks backend fallbacks because they force going
through IValue.

This PR removes that behavior and forces out arguments to be defined tensors.

It only looks at reduction ops for now, there's likely more PRs coming for other ops.

BC Breaking: This breaks BC of the C++ frontend API since those ops previously allowed calling with undefined tensors and that isn't allowed anymore.

Differential Revision: [D26795461](https://our.internmc.facebook.com/intern/diff/D26795461/)

[ghstack-poisoned]

smessmer committed:
Update on "Don't allocate result Tensors in out overloads: Reduction Ops"


We have some operators that previously allowed you to pass in an undefined tensor to the out argument,
and then would go on to allocate that for you. This behavior is broken and doesn't work in JIT when things
are converted to/from IValues. Because of this, it blocks backend fallbacks because they force going
through IValue.

This PR removes that behavior and forces out arguments to be defined tensors.

It only looks at reduction ops for now, there's likely more PRs coming for other ops.

BC Breaking: This breaks BC of the C++ frontend API since those ops previously allowed calling with undefined tensors and that isn't allowed anymore.

Differential Revision: [D26795461](https://our.internmc.facebook.com/intern/diff/D26795461/)

[ghstack-poisoned]

smessmer committed:
Update on "Don't allocate result Tensors in out overloads: Reduction Ops"


We have some operators that previously allowed you to pass in an undefined tensor to the out argument,
and then would go on to allocate that for you. This behavior is broken and doesn't work in JIT when things
are converted to/from IValues. Because of this, it blocks backend fallbacks because they force going
through IValue.

This PR removes that behavior and forces out arguments to be defined tensors.

It only looks at reduction ops for now, there's likely more PRs coming for other ops.

BC Breaking: This breaks BC of the C++ frontend API since those ops previously allowed calling with undefined tensors and that isn't allowed anymore.

Differential Revision: [D26795461](https://our.internmc.facebook.com/intern/diff/D26795461/)

[ghstack-poisoned]

smessmer committed:
Update on "Don't allocate result Tensors in out overloads: Reduction Ops"


We have some operators that previously allowed you to pass in an undefined tensor to the out argument,
and then would go on to allocate that for you. This behavior is broken and doesn't work in JIT when things
are converted to/from IValues. Because of this, it blocks backend fallbacks because they force going
through IValue.

This PR removes that behavior and forces out arguments to be defined tensors.

It only looks at reduction ops for now, there's likely more PRs coming for other ops.

BC Breaking: This breaks BC of the C++ frontend API since those ops previously allowed calling with undefined tensors and that isn't allowed anymore.

Differential Revision: [D26795461](https://our.internmc.facebook.com/intern/diff/D26795461/)

[ghstack-poisoned]

smessmer committed:
Update on "Don't allocate result Tensors in out overloads: Reduction Ops"


We have some operators that previously allowed you to pass in an undefined tensor to the out argument,
and then would go on to allocate that for you. This behavior is broken and doesn't work in JIT when things
are converted to/from IValues. Because of this, it blocks backend fallbacks because they force going
through IValue.

This PR removes that behavior and forces out arguments to be defined tensors.

It only looks at reduction ops for now, there's likely more PRs coming for other ops.

BC Breaking: This breaks BC of the C++ frontend API since those ops previously allowed calling with undefined tensors and that isn't allowed anymore.

Differential Revision: [D26795461](https://our.internmc.facebook.com/intern/diff/D26795461/)

[ghstack-poisoned]

smessmer committed:
Update on "Don't allocate result Tensors in out overloads: Reduction Ops"


We have some operators that previously allowed you to pass in an undefined tensor to the out argument,
and then would go on to allocate that for you. This behavior is broken and doesn't work in JIT when things
are converted to/from IValues. Because of this, it blocks backend fallbacks because they force going
through IValue.

This PR removes that behavior and forces out arguments to be defined tensors.

It only looks at reduction ops for now, there's likely more PRs coming for other ops.

BC Breaking: This breaks BC of the C++ frontend API since those ops previously allowed calling with undefined tensors and that isn't allowed anymore.

Differential Revision: [D26795461](https://our.internmc.facebook.com/intern/diff/D26795461/)

[ghstack-poisoned]
